### PR TITLE
feat!: enhance `Map` control

### LINKED ISSUE #3688: Enhance `Map` Control
- [x] programatic rotation
- [x] Zoom in and out
- [x] Move to specific points
- [x] cancellable tile provider

### PR DESCRIPTION
**Pull request by ndonkoHenri**
## Description
Closes #3688

## Code
<details><summary>Test Code</summary>
<p>

```python
import flet as ft
import flet.map as map
import random
import re

INITIAL_CENTER = map.MapLatitudeLongitude(latitude=17.701, longitude=11.44)
INITIAL_ZOOM = 4
TEXT_FIELD_ERROR_MSG = "Invalid input!"
TEXT_FIELD_INPUT_FILTER_REGEX = r"^\s*[+-]?\d*\.?\d*\s*$"


class MyTooltip(ft.Tooltip):
    """Custom tooltip with no wait duration and no preference to show above the control."""

    def __init__(self, text: str):
        super().__init__(text, wait_duration=0, prefer_below=False)


def main(page: ft.Page):
    page.padding = ft.Padding(2, 0, 2, 0)
    # on web, disable the browser context menu
    if page.web:
        page.browser_context_menu.disable()

    # refs
    marker_layer_ref = ft.Ref[map.MarkerLayer]()
    map_ref = ft.Ref[map.Map]()
    user_latitude_ref = ft.Ref[ft.TextField]()
    user_longitude_ref = ft.Ref[ft.TextField]()

    last_moved_to_marker_index = 0

    def handle_go_to_dlg_coordinates(e: ft.ControlEvent):
        lat = user_latitude_ref.current.value.rstrip(".")
        long = user_longitude_ref.current.value.rstrip(".")

        has_error = False

        # Validate latitude
        if not bool(re.match(r"[-+]?[0-9]*\.?[0-9]+", lat)):
            user_latitude_ref.current.error_text = TEXT_FIELD_ERROR_MSG
            has_error = True

        # Validate longitude
        if not bool(re.match(r"[-+]?[0-9]*\.?[0-9]+", long)):
            user_longitude_ref.current.error_text = TEXT_FIELD_ERROR_MSG
            has_error = True

        if not has_error:
            page.close(coordinates_dlg)
            map_ref.current.move_to(
                destination=map.MapLatitudeLongitude(
                    latitude=float(lat), longitude=float(long)
                ),
                zoom=INITIAL_ZOOM,
                animation_duration=ft.Duration(milliseconds=2000),
            )
            user_latitude_ref.current.error_text = None
            user_longitude_ref.current.error_text = None
        page.update()

    def handle_tap(e: map.MapTapEvent):
        """Handle tap event on the map. Adds a marker at the tapped location."""
        marker_layer_ref.current.markers.append(
            map.Marker(
                coordinates=e.coordinates,
                content=ft.GestureDetector(
                    mouse_cursor=ft.MouseCursor.CLICK,
                    content=ft.Icon(
                        name=ft.icons.LOCATION_ON,
                        size=68,
                        color=ft.cupertino_colors.BLACK,
                        tooltip=MyTooltip(
                            text=f"{round(e.coordinates.latitude, 3)}, {round(e.coordinates.longitude, 3)}"
                        ),
                    ),
                    on_tap=lambda _: map_ref.current.move_to(
                        destination=e.coordinates,
                        animation_duration=ft.Duration(milliseconds=2000),
                    ),
                ),
            )
        )
        page.update()

    def handle_clear_markers(e: ft.ControlEvent):
        """Clears all markers on the map."""
        nonlocal last_moved_to_marker_index
        marker_layer_ref.current.markers = []
        last_moved_to_marker_index = 0
        page.update()

    def handle_move_to_next_marker(e: ft.ControlEvent):
        """Moves the map to the next marker in the marker layer."""
        nonlocal last_moved_to_marker_index
        markers = marker_layer_ref.current.markers
        if len(markers) > 0:
            last_moved_to_marker_index = (last_moved_to_marker_index + 1) % len(markers)
            map_ref.current.move_to(
                destination=markers[last_moved_to_marker_index].coordinates,
                animation_duration=ft.Duration(milliseconds=2000),
            )
            page.update()

    coordinates_dlg = ft.AlertDialog(
        adaptive=True,
        title=ft.Text("Enter the coordinates"),
        content=ft.Column(
            tight=True,
            controls=[
                ft.TextField(
                    ref=user_latitude_ref,
                    label="Latitude",
                    input_filter=ft.InputFilter(
                        regex_string=TEXT_FIELD_INPUT_FILTER_REGEX
                    ),
                ),
                ft.TextField(
                    ref=user_longitude_ref,
                    label="Longitude",
                    input_filter=ft.InputFilter(
                        regex_string=TEXT_FIELD_INPUT_FILTER_REGEX
                    ),
                ),
            ],
        ),
        actions=[
            ft.TextButton(
                text="Cancel", on_click=lambda e: page.close(e.control.parent)
            ),
            ft.TextButton(text="GO", on_click=handle_go_to_dlg_coordinates),
        ],
    )
    page.add(
        ft.Stack(
            expand=True,
            controls=[
                map.Map(
                    ref=map_ref,
                    on_tap=handle_tap,
                    configuration=map.MapConfiguration(
                        initial_center=INITIAL_CENTER,
                        initial_zoom=INITIAL_ZOOM,
                        min_zoom=1.3,
                        animation_duration=ft.Duration(milliseconds=1500),
                        interaction_configuration=map.MapInteractionConfiguration(
                            flags=map.MapInteractiveFlag.ALL
                            & ~map.MapInteractiveFlag.DOUBLE_TAP_ZOOM  # prevent double tap zoom for quicker on-tap calls
                        ),
                    ),
                    layers=[
                        map.TileLayer(
                            url_template="https://tile.openstreetmap.org/{z}/{x}/{y}.png",
                            on_image_error=lambda e: print("TileLayer Error"),
                        ),
                        map.MarkerLayer(
                            ref=marker_layer_ref,
                            markers=[],
                        ),
                    ],
                ),
                ft.Row(
                    expand=True,
                    bottom=16,
                    left=40,
                    spacing=16,
                    alignment=ft.MainAxisAlignment.CENTER,
                    controls=[
                        ft.FloatingActionButton(
                            icon=ft.icons.ZOOM_IN,
                            tooltip=MyTooltip("Zoom in"),
                            on_click=lambda e: map_ref.current.zoom_in(
                                animation_duration=ft.Duration(milliseconds=2500)
                            ),
                        ),
                        ft.FloatingActionButton(
                            icon=ft.icons.ZOOM_OUT,
                            tooltip=MyTooltip("Zoom out"),
                            on_click=lambda e: map_ref.current.zoom_out(
                                animation_duration=ft.Duration(milliseconds=2500)
                            ),
                        ),
                        ft.FloatingActionButton(
                            icon=ft.icons.ROTATE_RIGHT,
                            tooltip=MyTooltip("Rotate 90°"),
                            on_click=lambda e: map_ref.current.rotate_from(degree=90),
                        ),
                        ft.FloatingActionButton(
                            icon=ft.icons.ROTATE_LEFT,
                            tooltip=MyTooltip("Rotate -90°"),
                            on_click=lambda e: map_ref.current.rotate_from(degree=-90),
                        ),
                        ft.FloatingActionButton(
                            icon=ft.icons.CROP_ROTATE,
                            tooltip=MyTooltip("Reset Rotation"),
                            on_click=lambda e: map_ref.current.reset_rotation(
                                animation_duration=ft.Duration(milliseconds=2000)
                            ),
                        ),
                        ft.FloatingActionButton(
                            icon=ft.icons.CENTER_FOCUS_STRONG,
                            tooltip=MyTooltip("Center"),
                            on_click=lambda e: map_ref.current.move_to(
                                destination=INITIAL_CENTER,
                                zoom=INITIAL_ZOOM,
                            ),
                        ),
                        ft.FloatingActionButton(
                            icon=ft.icons.EXPLORE,
                            tooltip=MyTooltip("Random Location"),
                            on_click=lambda e: map_ref.current.move_to(
                                destination=map.MapLatitudeLongitude(
                                    latitude=random.uniform(-90, 90),
                                    longitude=random.uniform(-180, 180),
                                ),
                                zoom=random.uniform(3, 6),
                                animation_duration=ft.Duration(milliseconds=4000),
                            ),
                        ),
                        ft.FloatingActionButton(
                            icon=ft.icons.POLYLINE_ROUNDED,
                            tooltip=MyTooltip("Move to next marker"),
                            on_click=handle_move_to_next_marker,
                        ),
                        ft.FloatingActionButton(
                            icon=ft.icons.SKIP_NEXT,
                            tooltip=MyTooltip("Move To"),
                            on_click=lambda e: page.open(coordinates_dlg),
                        ),
                        ft.FloatingActionButton(
                            icon=ft.icons.CLEAR,
                            tooltip=MyTooltip("Clear Markers"),
                            on_click=handle_clear_markers,
                        ),
                    ],
                ),
            ],
        )
    )


ft.app(main)

``` 

</p>
</details> 

## Summary by Sourcery

Enhance the Map control by adding new event handlers and methods for map interactions and animations. Refactor the MapConfiguration class and improve JSON parsing functions. Introduce animated map control and marker layers using new dependencies.

New Features:
- Introduce new event handlers for map interactions, including tap, hover, secondary tap, long press, and pointer events.
- Add new methods for map control such as rotate_from, reset_rotation, zoom_in, zoom_out, zoom_to, move_to, and center_on.
- Implement animated map control using the AnimatedMapController for smooth transitions and animations.
- Add support for animated markers and layers using the flutter_map_animations package.

Enhancements:
- Refactor the MapConfiguration class to use a dataclass for cleaner and more efficient attribute management.
- Improve the handling of optional values in JSON parsing functions by providing default values.
- Enhance the map's tile layer with a cancellable network tile provider and fade-in tile display.

Chores:
- Remove unused children parameters from various control classes to simplify the codebase.
- Update the pubspec.yaml to include new dependencies for map animations and tile provider.

### CODE REVIEW

File: sdk/python/packages/flet-core/src/flet_core/control.py [hunk 1]
```diff
@@ -337,6 +337,9 @@ def invoke_method(
         assert (
             self.__page
         ), f"{self.__class__.__qualname__} Control must be added to the page first"
+        if arguments:
+            # remove items with None values and convert other values to string
+            arguments = {k: str(v) for k, v in arguments.items() if v is not None}
         return self.__page._invoke_method(
```
Comment (FeodorFitsner) [inline sdk/python/packages/flet-core/src/flet_core/control.py:343]: What issue does it solve?

Comment (FeodorFitsner) [reply sdk/python/packages/flet-core/src/flet_core/control.py:343]: I'm not sure it won't break anything. I can be mistaken, but there could be some controls expecting bool or int on dart side from args JSON.

Comment (ndonkoHenri) [reply sdk/python/packages/flet-core/src/flet_core/control.py:343]: It's a while now, but i remember it was to always make sure values are always sent over as string (as required by definition: `arguments: Optional[Dict[str, str]]`). Kind of a pydantic-like validation process.
I had equally notice that sending `None` values to dart end lead to errors, so they are removed before being sent.

> I can be mistaken, but there could be some controls expecting bool or int on dart side from args JSON.

Checkout other controls like `Audio` to notice that we send everything as string no matter the type.
On dart end they are handled accordingly with utils like `parseDouble()`

Comment (Muddassir-Farooq-official) [reply sdk/python/packages/flet-core/src/flet_core/control.py:343]: Any update

Comment (ndonkoHenri) [reply sdk/python/packages/flet-core/src/flet_core/control.py:343]: @Muddassir-Farooq-official please stop spamming! It is really annoying. 
Simply be patient and give us time to work.

**Comment by FeodorFitsner** [comment]
Please resolve conflict.

**Review by FeodorFitsner — APPROVED** [review]

**Pull request merged**